File: library/src/main/java/io/blackbox_vision/mvphelpers/ui/view/BaseFrameLayout.java
package io.blackbox_vision.mvphelpers.ui.view;

import android.annotation.TargetApi;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import io.blackbox_vision.mvphelpers.logic.presenter.BasePresenter;


public abstract class BaseFrameLayout<T extends BasePresenter> extends FrameLayout {

    @Nullable
    private T presenter;

    public BaseFrameLayout(Context context) {
        super(context);
    }

    public BaseFrameLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public BaseFrameLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @TargetApi(21)
    public BaseFrameLayout(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        presenter = addPresenter();
        onPresenterCreated(presenter);
    }

    @Override
    protected void onDetachedFromWindow() {
        super.onDetachedFromWindow();

        if (null != presenter) {
            presenter.detachView();
        }
    }

    @NonNull
    public abstract T addPresenter();

    public abstract void onPresenterCreated(@NonNull T presenter);

    @Nullable
    public T getPresenter() {
        return presenter;
    }
}

File: library/src/main/java/io/blackbox_vision/mvphelpers/ui/view/BaseFrameLayout.kt
package io.blackbox_vision.mvphelpers.ui.view

import android.annotation.TargetApi
import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout

import io.blackbox_vision.mvphelpers.logic.presenter.BasePresenter


abstract class BaseFrameLayout<T : BasePresenter<*>> : FrameLayout {

    var presenter: T? = null
        private set

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    @TargetApi(21)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes)

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        presenter = addPresenter()
        onPresenterCreated(presenter!!)
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()

        if (null != presenter) {
            presenter!!.detachView()
        }
    }

    abstract fun addPresenter(): T

    abstract fun onPresenterCreated(presenter: T)
}
